### PR TITLE
Do not import duplicated modules with --importmode=importlib

### PR DESCRIPTION
**Pull request by nicoddemus**
Regression brought up by #11475.

### CODE REVIEW

**Comment by flying-sheep** [comment]
That failed import *is* one of the symptoms caused by this.

**Review by bluetech — APPROVED** [review]
Regardless of whether it fixes @flying-sheep's issue, the change LGTM

**Comment by flying-sheep** [comment]
Ah, sorry, the reproducer didn’t do what I thought it did.

This one works. I’ll check if this PR fixes things:

```py
git clone https://github.com/scverse/anndata.git
cd anndata
git checkout src2
python -m venv .venv
.venv/bin/pip install .[test] pytest@https://github.com/pytest-dev/pytest/releases/download/8.1.0/pytest-8.1.0-py3-none-any.whl
.venv/bin/pytest anndata src/anndata/tests/test_awkward.py::test_view src/anndata/tests/test_backed_sparse.py::test_empty_backed_indexing
```

**Comment by flying-sheep** [comment]
OK, I tried

```bash
git clone https://github.com/scverse/anndata.git
cd anndata
git checkout src2
python -m venv .venv
.venv/bin/pip install .[test] 'pytest@git+https://github.com/nicoddemus/pytest.git@11475-duplicated-imports'
.venv/bin/pytest anndata src/anndata/tests/test_awkward.py::test_view src/anndata/tests/test_backed_sparse.py::test_empty_backed_indexing
```

and it worked!

**Comment by nicoddemus** [comment]
Awesome, thanks @flying-sheep!

**Comment by aaraney** [comment]
I think this may have introduced a different regression. Consider the following project structure with two packages under the `ns` namespace:

```
ns
└── python
    ├── bar
    │   ├── ns
    │   │   ├── bar
    │   │   └── test
    │   │       ├── __init__.py
    │   │       ├── bar.py
    │   │       └── test_bar.py
    │   └── pyproject.toml
    └── foo
        ├── ns
        │   ├── foo
        │   └── test
        │       ├── __init__.py
        │       ├── foo.py
        │       └── test_foo.py
        └── pyproject.toml
```

In `pytest==8.0.2`, `python -m pytest --import-mode=importlib` correctly runs the tests from the top level `ns` directory. In `pytest==8.1.1`, `python -m pytest --import-mode=importlib -o "consider_namespace_packaes=true"`, results in the following error:

```
========================== test session starts ===========================
platform darwin -- Python 3.9.16, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/user/pytest-12074
collected 0 items / 2 errors

================================= ERRORS =================================
____________ ERROR collecting python/bar/ns/test/test_bar.py _____________
ImportError while importing test module '/home/user/pytest-12074/python/bar/ns/test/test_bar.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
python/bar/ns/test/test_bar.py:1: in <module>
    from .bar import value
E   ModuleNotFoundError: No module named 'test.bar'
____________ ERROR collecting python/foo/ns/test/test_foo.py _____________
ImportError while importing test module '/home/user/pytest-12074/python/foo/ns/test/test_foo.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
python/foo/ns/test/test_foo.py:1: in <module>
    from .foo import value
E   ModuleNotFoundError: No module named 'test.foo'
======================== short test summary info =========================
ERROR python/bar/ns/test/test_bar.py
ERROR python/foo/ns/test/test_foo.py
!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!
=========================== 2 errors in 0.04s ============================
```

**Comment by aaraney** [comment]
Just a little bit of a follow up, because I realize that my original comment wasn't the most clear. Below are the contents of `test_foo.py` and `foo.py`. `test_bar.py` and `bar.py` look nearly identical.

```
# python/foo/ns/test/test_foo.py
from .foo import value

def test_foo():
    assert value == "foo"
```

```
# python/foo/ns/test/foo.py
value = "foo"
```

**Comment by nicoddemus** [comment]
@aaraney thanks for the report, but could you open a new issue please (you can refer to this one there)?

I will take a look at it ASAP. :+1:

**Comment by aaraney** [comment]
@nicoddemus, for sure! Thanks for taking a look at this!

**Comment by tacaswell** [comment]
This appears to have broken assertion re-writing with editable installs.

**Comment by tacaswell** [comment]
This is fixed by #12313

File: src/_pytest/pathlib.py [hunk 1]
```diff
@@ -539,6 +539,10 @@ def import_path(
         except CouldNotResolvePathError:
             pass
         else:
+            # If the given module name is already in sys.modules, do not import it again.
+            with contextlib.suppress(KeyError):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/pathlib.py:543]: just taking a look again, i believe this should trigger a import mismatch if the file of the loaded module and the current path are different

**Pull request merged**